juju/bundle.py

```diff
@@ -659,6 +659,7 @@ class AddCharmChange(ChangeInfo):
     _toPy = {'charm': 'charm',
              'series': 'series',
              'channel': 'channel',
+             'revision': 'revision',
```

Comment:
I'm a little surprised that this is all it takes, though I guess the issue with the code that just follows this is that the `from_dict` base class looks for all the listed items and then will look to set the attributes based on the names that are given.

Reply:
I'm not that knowledgeable about the internals of python-libjuju but when debugging I saw that the revision was passed to this step as a parameter so to me it makes sense from a high-level view. also the other parameters like series and channel are also done in a similar manner.

It seems like I also need to update the unit tests, I'll find the time to work on this. I'll also add an integration test

Reply:
I fixed the issues with unit tests and also added an integration test for the pinned revision